R/merge_sce_list.R

```diff
@@ -263,3 +302,131 @@ prepare_sce_for_merge <- function(sce,
+#'  add those features in with `NA` counts, and update the rowData slot to match.
+#'  This involves creating a new SCE to replace the current altExp.
+#'
+#' @param sce The SCE object whose altExp should be prepared
+#' @param sce_name The name for this SCE
+#' @param altexp_features Vector of features that should be present in the altExp
+#'
+#' @return An updated SCE object with all altExp features present
+prepare_altexps_for_merge <- function(
+    sce,
+    sce_name,
+    altexp_features) {
+
+  sce_altexp <- altExp(sce)
+
+  # Determine which features are missing
+  missing_features <- altexp_features[!(altexp_features %in% rownames(sce_altexp))]
+  n_missing <- length(missing_features)
+
+  # Update altExp if any features are missing
+  if (n_missing > 0) {
+
+    # First, establish new rowData with all NA values
+    # this data.frame only contains rows for missing features
+    new_rowdata <- data.frame(
+      var = names(rowData(sce_altexp)),
+      val = rep(NA, n_missing)
+    ) |>
+      tidyr::pivot_wider(
+        names_from = var,
+        values_from = val
+      ) |>
+      # repeat rows n_missing times
+      dplyr::slice(
+        rep(
+          1:dplyr::n(),
+          each = n_missing
+        )
+      )
+
+    # create new rownames to be added in when we re-DataFrame this
+    new_rownames <- c(
+      rownames(sce_altexp), # existing rownames
+      missing_features # new rownames
+    )
+
+    # Next, establish new counts and logcounts assays
+    # TODO: Can either add `if` for whether to run on each assay, or fail if
+    #  either is missing. Any reviewer preferences?
+    new_counts <- update_altexp_assay(
+      counts(sce_altexp),
+      missing_features
+    )
+
+    new_logcounts <- update_altexp_assay(
+      logcounts(sce_altexp),
+      missing_features
+    )
+
+    # Create a new altexp and add it into the SCE
+    # TODO: this line needs to be changed if we want to do `if` with assay update
+    new_altexp <- SingleCellExperiment(assays = list(counts = new_counts,
+                                                     logcounts = new_logcounts))
+
+    # Add the new rowData rows into the SCE, with updated column names
+    rowData(new_altexp) <- sce_altexp |>
+      rowData() |>
+      as.data.frame() |>
+      rbind(new_rowdata) |>
+      # re DataFrame with _all_ feature rownames
+      DataFrame(row.names = new_rownames)
+
+    # Add colData and metadata
+    colData(new_altexp) <- colData(altExp(sce))
+    metadata(new_altexp) <- metadata(altExp(sce))
+
+    # Replace the old altExp, while preparing it for merge
+    altExp(sce) <- prepare_sce_for_merge(
+      new_altexp,
+      sce_name,
+      batch_column = "library_id",
+      cell_id_column = "cell_id",
+      shared_features = altexp_features,
+      retain_coldata_cols = colnames(colData(new_altexp)),
+      preserve_rowdata_cols = NA # rename them all
+    )
+  }
```

Comment:
this doesn't quite work. You haven't accounted for preparing the SCE if there are no missing features.